src/Background/NotificationHandler.vala

```diff
@@ -0,0 +1,109 @@
+            HashTable<string, Variant> hints,
+            int32 expire_timeout
+        ) throws DBusError, IOError;
+    }
+
+    public NotificationHandler (DBusConnection connection) {
+        this.connection = connection;
+        notification_by_id = new HashTable<uint32, NotificationRequest> (null, null);
+
+        try {
+            notifications = connection.get_proxy_sync ("org.freedesktop.Notifications", "/org/freedesktop/Notifications");
+            notifications.action_invoked.connect (on_action_invoked);
+            notifications.notification_closed.connect (on_notification_closed);
+        } catch {
+            warning ("Cannot connect to notifications dbus, background portal working with reduced functionality.");
+        }
+    }
+
+    public NotificationRequest? send_notification (string app_id, string app_name) {
+        string[] actions = {
+            ACTION_ALLOW_BACKGROUND,
+            _("Allow"),
+            ACTION_FORBID_BACKGROUND,
+            _("Forbid")
+        };
+
+        try {
+            var id = notifications.notify (
+                Environment.get_prgname (),
+                0,
+                "dialog-information",
+                _("Background activity"),
+                _(""""%s" is running in the background""").printf (app_name),
+                actions,
+                new HashTable<string, Variant> (str_hash, str_equal),
+                0
+            );
+
+            var notification = new NotificationRequest (this, id);
+            notification_by_id.set (id, notification);
+
+            return notification;
+        } catch (Error e) {
+            warning ("Failed to send notification for app id '%s': %s", app_id, e.message);
+            return null;
+        }
+    }
+
+    private void on_action_invoked (uint32 id, string action_key) {
+        if (id in notification_by_id) {
+            var notification = notification_by_id.get (id);
+            if (action_key == NotificationHandler.ACTION_ALLOW_BACKGROUND) {
+                notification.response (NotifyBackgroundResult.ALLOW);
+            } else if (action_key == NotificationHandler.ACTION_FORBID_BACKGROUND) {
+                notification.response (NotifyBackgroundResult.FORBID);
+            }
+
+            notification_by_id.remove (id);
+        }
+    }
+
+    private void on_notification_closed (uint32 id, uint32 reason) {
+        if (id in notification_by_id) {
+            var notification = notification_by_id.get (id);
+            if (reason == 2 || reason == 3) {
+                notification.response (NotifyBackgroundResult.ALLOW_ONCE);
+
+                notification_by_id.remove (id);
+            }
+        }
+    }
+
+    public void close_notification (uint32 id) {
+        try {
+            notifications.close_notification (id);
+        } catch (Error e) {
+            warning ("Failed to close notification with id '%u': %s", id, e.message);
+        }
+
+        notification_by_id.remove (id);
```

Comment:
There's two ways this can error out, by the notification not existing in the server, or by the server not existing. so don't need to print the error in this case.

in the success case, the notification_closed signal will already handle removing the notification from the table so we don't need to do it here.

```suggestion
        try {
            notifications.close_notification (id);
        } catch {
           // the notification was already closed, or we lost the connection to the server
        }
```

Reply:
I think if it errors out we probably still should remove the notification manually because the closed signal might not be send then right?

Reply:
i worry that we end leaking the notification in the server, if the call to remove it failed, it still exists, but we will end ignoring any action made by the user at this point.

Reply:
I think any interaction made by the user after a close call will be ignored anyway because I can't imagine the front-end will still listen to the request object after it made a close call. But I guess it also won't hurt to keep the request open until user interaction and then just silently fail 🤷